File: Proiectarea Algoritmilor/Lab05/codeBase/C++/src-lab5/lab5p1.cpp
#include <iostream>
#include <fstream>
#include <vector>

#include "XOBoard.h"

#define INF 8

#define OPPONENT(x) ((x) == XOBoard::PlayerX ? \
                     XOBoard::PlayerO : XOBoard::PlayerX)

/** Un exemplu de functie de gandire care foloseste rezultatele furnizate de
 * o abordare negaMax exhaustiva. */
XOBoard negaMaxThink(XOBoard::Player player, XOBoard board)
{
  /* TODO: Fa o miscare si returneaza tabla dupa aceasta miscare. Aceasta
   * functie de AI trebuie sa respecte acest format pentru ca este data in
   * constructorul jocului, dar puteti apela aici o functie scrisa de voi. */

  /* TODO: Stergeti linia de mai jos dupa ce rezolvati. */
  return board;
}

int main()
{
  /** Cream un engine de joc si specificam ca am vrea ca AI-ul sa gandeasca
   * cu functia data ca parametru (o functie de tipul:
   *       XOBoard ComputerAI(XOBoard::Player, XOBoard)).
   *
   * Engine-ul de joc va apela intern aceasta functie atunci cand pune
   * calculatorul sa gandeasca. Parametrii au urmatoarea semnificatie:
   *
   * player = jucatorul care trebuie sa mute in continuare (identitatea
   *          calculatorului care gandeste cu aceasta functie).
   *           Valorile posibile sunt { XOBoard::PlayerX, XOBoard::PlayerO }
   *
   * board = tabla pe care o vede jucatorul care trebuie sa mute in
   *         continuare.
   */
  XOGame game(negaMaxThink);

  /** Lansam un joc intre un om (joaca cu X) si un calculator (joaca cu O) 
   * Puteti lansa si computerVShuman, humanVShuman, dar si computerVScomputer.
   * Mai multe detalii in documentatie.
   * (codeBase/C++/doc/html/index.html -> classes)
   */
  game.humanVScomputer();

  return 0;
}

File: Proiectarea Algoritmilor/Lab05/codeBase/Java/src-lab5/Lab5p1.java

import java.util.Vector;

public class Lab5p1 {

    public static void main(String[] args) {
        XOGame game = new XOGame(new SmartComputerAI(), new Human());
        game.runGame();
    }
}

class ScoreBoardPair {

    XOBoard board;
    int score;
    int steps;

    public ScoreBoardPair(int score, XOBoard board, int steps) {
        this.score = score;
        this.board = board;
        this.steps = steps;
    }
}

class SmartComputerAI implements IsAbleToPlay {

    private static final int INF = 100;

    private XOPlayer opponent(XOPlayer player) {
        if (player == XOPlayer.PlayerO) {
            return XOPlayer.PlayerX;
        } else {
            return XOPlayer.PlayerO;
        }
    }

    public ScoreBoardPair best_move(int alfa,int beta, XOPlayer player, XOBoard board, int steps) {
        XOBoard b = null;
        int score = -INF;
        steps = 10000;
        if (board.get_score(player) > board.get_score(opponent(player)))
            return new ScoreBoardPair(board.get_score(player) - board.get_score(opponent(player)), board, steps);
        if (board.is_full()) {
            return new ScoreBoardPair(board.get_score(player) - board.get_score(opponent(player)), board, steps);
        } else {
            ScoreBoardPair x;
            for (int i = 0; i < 3; i++) {
                for (int j = 0; j < 3; j++) {
                    if (board.is_empty(i, j)) {
                        board.put(player, i, j);
                        x = best_move(-beta, -alfa, opponent(player), board, steps + 1);
                        if (-x.score >= beta)
                        {
                              b = board.tryClone();
                              b.erase(i, j);
                              return new ScoreBoardPair(beta, b, steps);
                        }
                        if (-x.score > alfa || (-x.score == alfa && x.steps < steps)) {
                            b = board.tryClone();
                            alfa = -x.score;
                            steps = x.steps;
                        }
                        board.erase(i, j);
                    }
                }
            }
            return new ScoreBoardPair(alfa, b, steps);
        }
    }

    public XOBoard move(XOPlayer player, XOBoard board) {
        System.out.println("Computer is thinking...");
        /*
         * TODO: Fa o miscare si returneaza tabla dupa aceasta miscare. Aceasta
         * functie de AI trebuie sa respecte acest format pentru ca este data in
         * constructorul jocului, dar puteti apela aici o functie scrisa de voi.
         */
        return best_move(INF, -INF, player, board, 0).board;
        /*
         * TODO: Stergeti linia de mai jos dupa ce rezolvati.
         */
    }
}
